File: apps/api/app/schemas/jobs.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.entities import JobStatus
from app.models.job_stages import JobStage


class IndexingJobRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    repository_id: UUID
    snapshot_id: UUID | None
    status: JobStatus
    stage: str
    progress_percentage: int = Field(ge=0, le=100)
    attempt_count: int = Field(ge=0)
    max_attempts: int = Field(ge=1)
    locked_by: str | None
    locked_until: datetime | None
    heartbeat_at: datetime | None
    error_code: str | None
    error_message: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class IndexingJobCreateDefaults(BaseModel):
    """Defaults applied when enqueueing a new indexing job."""

    status: JobStatus = JobStatus.QUEUED
    stage: JobStage = JobStage.QUEUED
    progress_percentage: int = 0
    attempt_count: int = 0
    max_attempts: int = 3

File: apps/web/src/lib/jobs.ts
export type JobStatus = "QUEUED" | "RUNNING" | "SUCCEEDED" | "FAILED" | "CANCELLED";

export type JobStage =
  | "queued"
  | "cloning"
  | "discovering_files"
  | "parsing"
  | "building_relationships"
  | "chunking"
  | "embedding"
  | "validating"
  | "completed";

export interface IndexingJob {
  id: string;
  repository_id: string;
  snapshot_id: string | null;
  status: JobStatus;
  stage: JobStage | string;
  progress_percentage: number;
  attempt_count: number;
  max_attempts: number;
  locked_by: string | null;
  locked_until: string | null;
  heartbeat_at: string | null;
  error_code: string | null;
  error_message: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface RepositorySummary {
  id: string;
  host: string;
  owner_name: string;
  name: string;
  default_branch: string;
  clone_url: string;
}

export interface RepositoryImportResponse {
  repository: RepositorySummary;
  job: IndexingJob;
  created_new_job: boolean;
}

export const JOB_STAGES: Array<{ id: JobStage; label: string }> = [
  { id: "queued", label: "Queued" },
  { id: "cloning", label: "Cloning" },
  { id: "discovering_files", label: "Discovering files" },
  { id: "parsing", label: "Parsing" },
  { id: "building_relationships", label: "Building relationships" },
  { id: "chunking", label: "Chunking" },
  { id: "embedding", label: "Embedding" },
  { id: "validating", label: "Validating" },
  { id: "completed", label: "Completed" },
];

export function stageIndex(stage: string): number {
  const index = JOB_STAGES.findIndex((item) => item.id === stage);
  return index >= 0 ? index : 0;
}

export function isTerminalStatus(status: JobStatus): boolean {
  return status === "SUCCEEDED" || status === "FAILED" || status === "CANCELLED";
}
